c/parallel/src/for.cu

```diff
@@ -186,3 +188,30 @@ CUresult cccl_device_for_cleanup(cccl_device_for_build_result_t* bld_ptr)
 
   return CUDA_SUCCESS;
 }
+
+} // namespace for_impl
+
+CUresult cccl_device_for_build(
+  cccl_device_for_build_result_t* build_ptr,
+  cccl_iterator_t d_data,
+  cccl_op_t op,
+  int cc_major,
+  int cc_minor,
+  const char* cub_path,
+  const char* thrust_path,
+  const char* libcudacxx_path,
+  const char* ctk_path)
+{
+  return for_impl::build(build_ptr, d_data, op, cc_major, cc_minor, cub_path, thrust_path, libcudacxx_path, ctk_path);
```

Comment:
Q: Why do we need a second function in a namespace and not just put the implemention here?

Reply:
In short, because we wanted the implementation function to be `noexcept`. 

My train of though was as follows.

 The `noexcept` is C++ only keyword so it can be present in the header if we want it to be parseable by a C compiler.  C++ compile errors if the declaration does not have `noexcept` but the implementation does.  Removing `noexcept` from C++ function would mean that compiler generates additional logic for handling possible exceptions. The indirection allows to work around this. Hopefully, compiler is smart enough to see that even though the function, say `cccl_device_reduce` is not declared `noexcept`, it only calls a `noexcept` implementation which should get inline.

Reply:
> In short, because we wanted the implementation function to be noexcept.

That's what I assumed but this is not a compelling reason IMO. The value of `noexcept` diminishes very quickly for larger functions. `noexcept` can also add overhead, if the compiler cannot perform a full analysis of the body, since you must catch any exeception and call `terminate()`.

But all this is speculation without looking at the disassembly. However, this handling adds a fair amount of LOCs to the code base. I think it's not worth the cost, but it's also not a blocker.

Reply:
Let me know if you'd like me to restore it like before except without `noexcept`? It is fast to do.

Reply:
I would do it, but it's your c-parallel is not my domain :) So your choice!

Reply:
Change pushed